sql/parser/type_check.go

```diff
@@ -177,67 +176,44 @@ func typeCheckComparisonExpr(expr *ComparisonExpr) (Datum, error) {
 		return nil, err
 	}
 
-	return typeCheckComparisonOp(expr.Operator, leftType, rightType)
+	d, cmp, err := typeCheckComparisonOp(expr.Operator, leftType, rightType)
+	expr.fn = cmp
+	return d, err
 }
 
-func typeCheckComparisonOp(op ComparisonOp, dummyLeft, dummyRight Datum) (Datum, error) {
+func typeCheckComparisonOp(op ComparisonOp, dummyLeft, dummyRight Datum) (Datum, cmpOp, error) {
+	op, dummyLeft, dummyRight, _ = foldComparisonExpr(op, dummyLeft, dummyRight)
+
 	if dummyLeft == DNull || dummyRight == DNull {
 		switch op {
 		case Is, IsNot, IsDistinctFrom, IsNotDistinctFrom:
 			// TODO(pmattis): For IS {UNKNOWN,TRUE,FALSE} we should be requiring that
 			// dummyLeft == DummyBool. We currently can't distinguish NULL from
 			// UNKNOWN. Is it important to do so?
-			return DummyBool, nil
+			return DummyBool, cmpOp{}, nil
```

Comment:
Yeah good catch. I've moved the `foldComparisonExpr` transform to after this special case and addressed the need for it in the comments for `Is/IsNot` and `IsDistinctFrom/IsNotDistinctFrom`.